File: web/lib/publish/upload-media.ts
import {
  registerLinkedInUpload,
  uploadBytesToLinkedIn,
} from "@/lib/adapters/linkedin";
import { uploadMediaToX } from "@/lib/adapters/x";

export function mimeTypeFromUrl(url: string): string {
  const ext = url.split("?")[0].split(".").pop()?.toLowerCase();
  switch (ext) {
    case "png": return "image/png";
    case "gif": return "image/gif";
    case "webp": return "image/webp";
    case "jpg":
    case "jpeg":
    default: return "image/jpeg";
  }
}

// Fetches image bytes from a Cloudinary URL.
// Returns null on failure (non-fatal — caller publishes text-only).
async function fetchImageBytes(cloudinaryUrl: string): Promise<Buffer | null> {
  try {
    const res = await fetch(cloudinaryUrl);
    if (!res.ok) return null;
    return Buffer.from(await res.arrayBuffer());
  } catch {
    return null;
  }
}

export type MediaForPublish = {
  cloudinary_url: string;
};

// Uploads each media asset to the target platform.
// Returns platform-specific IDs/URNs in the same order as input.
// Failures are non-fatal — logs a warning and returns empty array.
export async function uploadMediaForPlatform(
  platform: "linkedin" | "x",
  accessToken: string,
  assets: MediaForPublish[],
  authorUrn?: string // LinkedIn only
): Promise<string[]> {
  if (assets.length === 0) return [];

  const ids: string[] = [];

  for (const asset of assets) {
    try {
      const mimeType = mimeTypeFromUrl(asset.cloudinary_url);
      const imageBytes = await fetchImageBytes(asset.cloudinary_url);
      if (!imageBytes) {
        console.warn(`[upload-media] Failed to fetch bytes from ${asset.cloudinary_url} — skipping`);
        continue;
      }

      if (platform === "linkedin") {
        if (!authorUrn) throw new Error("authorUrn required for LinkedIn");
        const { uploadUrl, assetUrn } = await registerLinkedInUpload(
          accessToken,
          authorUrn,
          imageBytes.length
        );
        await uploadBytesToLinkedIn(uploadUrl, imageBytes);
        ids.push(assetUrn);
      } else {
        const mediaId = await uploadMediaToX(accessToken, imageBytes, mimeType);
        ids.push(mediaId);
      }
    } catch (err) {
      console.warn(
        `[upload-media] Media upload failed for ${asset.cloudinary_url}:`,
        err instanceof Error ? err.message : err
      );
      // Non-fatal: skip this asset, continue with others
    }
  }

  return ids;
}

File: worker/publish/upload_media.py
import httpx
import structlog

from adapters.platforms import get_adapter

log = structlog.get_logger()


def mime_type_from_url(url: str) -> str:
    ext = url.split("?")[0].split(".")[-1].lower()
    return {
        "png": "image/png",
        "gif": "image/gif",
        "webp": "image/webp",
        "jpg": "image/jpeg",
        "jpeg": "image/jpeg",
    }.get(ext, "image/jpeg")


async def _fetch_image_bytes(url: str) -> bytes | None:
    """Returns None on failure (non-fatal — caller publishes text-only)."""
    try:
        async with httpx.AsyncClient(timeout=httpx.Timeout(30.0)) as client:
            res = await client.get(url)
        if res.status_code >= 400:
            return None
        return res.content
    except Exception:  # noqa: BLE001 — non-fatal
        return None


async def upload_media_for_platform(
    platform: str,
    access_token: str,
    cloudinary_urls: list[str],
    author_urn: str | None = None,
) -> list[str]:
    """Uploads each asset to the target platform, returning platform media
    IDs/URNs. Failures are non-fatal — a failed asset is skipped."""
    ids: list[str] = []
    adapter = get_adapter(platform)
    for url in cloudinary_urls:
        try:
            mime = mime_type_from_url(url)
            data = await _fetch_image_bytes(url)
            if not data:
                log.warning("media_fetch_failed", url=url)
                continue
            ids.append(await adapter.upload_media(access_token, data, mime, author_urn))
        except Exception as exc:  # noqa: BLE001 — non-fatal, skip this asset
            log.warning("media_upload_failed", url=url, error=str(exc))
    return ids
